net_transport.go

```diff
@@ -155,6 +161,63 @@ func NewNetworkTransportWithLogger(
 	return trans
 }
 
+// NewNetworkTransportWithServerAddressProvider creates a new network transport with the given dialer
```

Comment:
This pattern won't scale super well and duplicates a lot. How about a `NewNetworkTransportWithConfig` with a new `NetworkTransportConfig` struct that has an optional provider, and then you can call that from the old `NewNetworkTransport` and `NewNetworkTransportWithLogger` methods as well?